Altimeter/Src/altimeter.cpp

```diff
@@ -0,0 +1,166 @@
+#include <altimeter.hpp>
+
+
+AltDevice::AltDevice(GPIO_TypeDef *ps_port, uint16_t ps_pin){
+	HAL_GPIO_WritePin(ps_port, ps_pin, GPIO_PIN_RESET);
+}
+
+void AltDevice::altInit(SPI_HandleTypeDef *spi_handle, GPIO_TypeDef *cs_port, uint16_t cs_pin){
+	reset(spi_handle, cs_port, cs_pin);
+	getConvCoeffs(spi_handle, cs_port, cs_pin);
+
+	/*Get the average of 100 readings*/
+
+	float reading_sum = 0.0f;
+	for(int i = 0; i < 100; i++){
+		calculateTempPres(spi_handle, cs_port, cs_pin);
+		reading_sum += getAltAboveSeaLvl();
+	}
+
+	base_elev_ = reading_sum / 100.0f;
+}
+
+void AltDevice::reset(SPI_HandleTypeDef *spi_handle, GPIO_TypeDef *cs_port, uint16_t cs_pin){
+
+	HAL_GPIO_WritePin(cs_port, cs_pin, GPIO_PIN_SET);
+	HAL_GPIO_WritePin(cs_port, cs_pin, GPIO_PIN_RESET);
+
+	uint8_t send_reset[3] = {RESET, 0xFF, 0xFF};
+	uint8_t receive_data[3];
+
+	HAL_SPI_TransmitReceive(spi_handle, send_reset, receive_data, 3, TIMEOUT);
+	osDelay(3);
+
+	HAL_GPIO_WritePin(cs_port, cs_pin, GPIO_PIN_SET);
+}
+
+void AltDevice::getConvCoeffs(SPI_HandleTypeDef *spi_handle, GPIO_TypeDef *cs_port, uint16_t cs_pin){
+	coeffs_.sens = promRead(spi_handle, cs_port, cs_pin, PROM_READ_ADDRESS_1);
+	coeffs_.off = promRead(spi_handle, cs_port, cs_pin, PROM_READ_ADDRESS_2);
+	coeffs_.tcs = promRead(spi_handle, cs_port, cs_pin, PROM_READ_ADDRESS_3);
+	coeffs_.tco = promRead(spi_handle, cs_port, cs_pin, PROM_READ_ADDRESS_4);
+	coeffs_.t_ref = promRead(spi_handle, cs_port, cs_pin, PROM_READ_ADDRESS_5);
+	coeffs_.temp_sens = promRead(spi_handle, cs_port, cs_pin, PROM_READ_ADDRESS_6);
+}
+
+
+uint16_t AltDevice::promRead(SPI_HandleTypeDef *spi_handle, GPIO_TypeDef *cs_port, uint16_t cs_pin, uint8_t address){
+	HAL_GPIO_WritePin(cs_port, cs_pin, GPIO_PIN_SET);
+	HAL_GPIO_WritePin(cs_port, cs_pin, GPIO_PIN_RESET);
+
+	uint8_t send_read_command[3] = {address, 0xFF, 0xFF};
+	uint8_t coefficient_val[3];
+
+	HAL_SPI_TransmitReceive(spi_handle, send_read_command, coefficient_val, 3, TIMEOUT);
+
+	HAL_GPIO_WritePin(cs_port, cs_pin, GPIO_PIN_SET);
+
+	//Grab conversion coefficients.
+	uint16_t conversion_coefficient = (coefficient_val[1] << 8) | (coefficient_val[2]);
+
+	return conversion_coefficient;
+}
+
+
+
+uint32_t AltDevice::uncompensatedPressureTemperature(SPI_HandleTypeDef *spi_handle, GPIO_TypeDef *cs_port, uint16_t cs_pin, uint8_t conversion_command){
+
+	uint8_t digital_pres_temp[1] = {conversion_command};
+	uint8_t receive_buffer[1];
+
+	HAL_GPIO_WritePin(cs_port, cs_pin, GPIO_PIN_SET);
+	HAL_GPIO_WritePin(cs_port, cs_pin, GPIO_PIN_RESET);
+
+	/* send the digital pressure or temperature with the command */
+	HAL_SPI_TransmitReceive(spi_handle, digital_pres_temp, receive_buffer, 1, TIMEOUT);
```

Comment:
I think this function could be changed to an interrupt based version of SPI receive in the next revision, so when the worst situation happened, we are not waiting for this timeout forever. Every return value from the communication function call to an external device should be checked. The failed communication should be handled explicitly, and inform to the function caller.

Reply:
Okay